buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -0,0 +1,416 @@
+    traverse(queue, rootDirectory, functionMap);
+    return functionMap;
+  }
+
+  /**
+   * Generates the parserImpls.ftl file for the dialect.
+   *
+   * @param functions The functions to place into the output file
+   */
+  public void generateParserImpls(Map<String, String> functions) {
+    // TODO(AndrewPochapsky): Add generation logic.
+  }
+
+  /**
+   * Gets the traversal path for the dialect by "subtracting" the root
+   * absolute path from the dialect directory absolute path.
+   *
+   * @param rootDirectoryFile The file for the root parsing directory
+   */
+  private Queue<String> getTraversalPath(File rootDirectoryFile) {
+    String dialectPath = dialectDirectory.getAbsolutePath();
+    String rootPath = rootDirectory.getAbsolutePath();
+    int rootIndex = dialectPath.indexOf(rootPath);
+    dialectPath = dialectPath.substring(rootIndex + rootPath.length() + 1);
+    return new LinkedList(Arrays.asList(dialectPath.split("/")));
+  }
+
+  /**
+   * Traverses the determined path given by the queue. Once the queue is
+   * empty, the dialect directory has been reached. In that case any *.ftl
+   * file should be processed and no further traversal should happen.
+   *
+   * @param directories The directories to traverse in topdown order
+   * @param currentDirectory The current directory the function is processing
+   * @param functionMap The map to which the parsing functions will be added to
+   */
+  private void traverse(
+      Queue<String> directories,
+      File currentDirectory,
+      Map<String, String> functionMap) {
+    File[] files = currentDirectory.listFiles();
+    // Ensures that files are processed first.
+    Arrays.sort(files, fileComparator);
+    String nextDirectory = directories.peek();
+    for (File f : files) {
+      String extension = "";
+      String fileName = f.getName();
+      int i = fileName.lastIndexOf('.');
+      if (i > 0) {
+        extension = fileName.substring(i+1);
+      }
+      if (f.isFile() && extension.equals("ftl")) {
+        try {
+          processFile(f, functionMap);
+        } catch(IOException e) {
+          e.printStackTrace();
+        }
+      } else if (!directories.isEmpty() && fileName.equals(nextDirectory)) {
+        // Remove the front element in the queue, the value is used above with
+        // directories.peek().
+        directories.poll();
+        traverse(directories, f, functionMap);
+      }
+    }
+  }
+
+  /**
+   * Extracts the functions from the given file into functionMap. Parses
+   * functions of the form:
+   * <return_type> <name>(<args>) :
+   * {
+   *     <properties>
+   * }
+   * {
+   *     <body>
+   * }
+   *
+   * @param file The file to process
+   * @param functionMap The map to which the parsing functions will be added to
+   */
+  private void processFile(File file, Map<String, String> functionMap) throws IOException {
+    String fileText = new String(Files.readAllBytes(Paths.get(file.getAbsolutePath())),
```

Comment:
We should try to extract the Files.readAllBytes() out of this function and have this function instead just take the file contents as input.

That would make it a lot more straightforward to unit-test this function without having to do file I/O.